File: TypeScript/codility-coding-excercises/src/lessons/lesson4.1.scala
package antivanov.codility.lessons

case class CrossingState(waitingForLeavesToFall: Set[Int], earliestTimeToCross: Option[Int] = None)

def earliestMomentToCross(riverLength: Int, fallingLeavesInTime: Seq[Int]): Int = {
  val leavesToFall = (1 to riverLength).toSet
  val finalCrossingState: CrossingState = fallingLeavesInTime.zipWithIndex.foldLeft(CrossingState(leavesToFall))({
    case (CrossingState(leftLeaves, Some(timeToCross)), _) => CrossingState(leftLeaves, Some(timeToCross))
    case (CrossingState(leftLeaves, None), (leafPosition, currentTime)) =>
      val updatedLeavesLeftToFall = leftLeaves - leafPosition
      val timeToCross = if (updatedLeavesLeftToFall.isEmpty)
        Some(currentTime)
      else
        None
      CrossingState(updatedLeavesLeftToFall, timeToCross)
  })
  finalCrossingState.earliestTimeToCross.getOrElse(-1)
}

@main def lesson41Main: Unit =
  val riverLength = 5
  val fallingLeavesInTime = Seq(1, 3, 1, 4, 2, 3, 5, 4)
  assert(earliestMomentToCross(riverLength, fallingLeavesInTime) == 6)

File: TypeScript/codility-coding-excercises/src/lessons/lesson4.1.ts
//TODO: Carefully check corner cases of riverLength and faillintLeavesInTime
function earliestMomentToCross(riverLength: number, fallingLeavesInTime: Array<number>): number {
  let leavesToFall = new Set(Array.from(Array(riverLength).keys()))
  let currentMoment = 0
  while (currentMoment < fallingLeavesInTime.length && leavesToFall.size > 0) {
    leavesToFall.delete(fallingLeavesInTime[currentMoment] - 1)
    currentMoment++
  }
  if (leavesToFall.size == 0) {
    return currentMoment - 1
  } else {
    return -1
  }
}

console.log(earliestMomentToCross(5, [1, 3, 1, 4, 2, 3, 5, 4]))